### PR TITLE
Implement the contao:crawl command

### PR DESCRIPTION
**Pull request by Toflar**
This is a sub PR of https://github.com/contao/contao/pull/981 only containing the CLI variant. It will allow us to focus more on the technical implementation first before we have all UI/CSS changes.

### CODE REVIEW

**Comment by Toflar** [comment]
This needs another update for latest Escargot changes. Will work on them tomorrow.

File: core-bundle/src/DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -343,6 +344,40 @@ private function addSearchNode(): NodeDefinition
...
+            ->getRootNode()
+            ->addDefaultsIfNotSet()
+            ->children()
+                ->arrayNode('additionalURIs')
+                    ->info('Additional URIs to crawl (by default, only the ones defined in the root pages are crawled).')
+                    ->validate()
+                    ->ifTrue(
+                        static function (array $uris): bool {
+                            foreach ($uris as $uri) {
+                                if (!preg_match('@^https?://@', $uri)) {
```
Comment (leofeyer) [inline core-bundle/src/DependencyInjection/Configuration.php:373]: Is this correct? The function returns `true` if the URL does **not** begin with `http://` or `https://`.

File: core-bundle/src/DependencyInjection/ContaoCoreExtension.php [hunk 2]
```diff
@@ -95,6 +97,11 @@ public function load(array $configs, ContainerBuilder $container): void
             ->registerForAutoconfiguration(PickerProviderInterface::class)
             ->addTag('contao.picker_provider')
         ;
+
+        $container
+            ->registerForAutoconfiguration(EscargotSubscriber::class)
+            ->addTag('contao.escargot_subscriber')
+        ;
```
Comment (leofeyer) [inline core-bundle/src/DependencyInjection/ContaoCoreExtension.php]: Did you intentionally not put this inside the `handleCrawlConfig()` method?

File: core-bundle/src/EventListener/SearchIndexListener.php [hunk 3]
```diff
@@ -60,10 +61,14 @@ public function onKernelTerminate(TerminateEvent $event): void
...
-        } else {
-            $this->indexer->delete($document);
+        try {
+            if ($event->getResponse()->isSuccessful()) {
+                $this->indexer->index($document);
+            } else {
+                $this->indexer->delete($document);
+            }
+        } catch (IndexerException $e) {
+            // Noop
```
Comment (leofeyer) [inline core-bundle/src/EventListener/SearchIndexListener.php]: Do we not want this to be logged?

File: core-bundle/tests/DependencyInjection/ContaoCoreExtensionTest.php [hunk 4]
```diff
@@ -1884,6 +1884,40 @@ public function testRegistersTheTwigTemplateExtension(): void
...
+                ],
+            ],
+            $this->container
+        );
+
+        $factory = $this->container->getDefinition('contao.search.escargot_factory');
+
+        $this->assertSame(['https://examle.com'], $factory->getArgument(2));
+        $this->assertSame(['proxy' => 'http://localhost:7080'], $factory->getArgument(3));
+    }
```
Comment (leofeyer) [inline core-bundle/tests/DependencyInjection/ContaoCoreExtensionTest.php:1979]: We also have to test the `contao.search.escargot_subscriber.search_index` service here, don't we?

File: core-bundle/src/DependencyInjection/Configuration.php (hunk 1 above)
Comment (Toflar) [reply core-bundle/src/DependencyInjection/Configuration.php:373]: Yes, this is correct. `ifTrue` -> `thenFail`. It's even covered by unit tests ;)

File: core-bundle/src/DependencyInjection/ContaoCoreExtension.php (hunk 2 above)
Comment (Toflar) [reply core-bundle/src/DependencyInjection/ContaoCoreExtension.php]: Ah, no. It was just where the other registerForAutoconfiguration call was :) Fixed in 63b0fdfcadb813f2d10191996dcca04909febd95.

File: core-bundle/src/EventListener/SearchIndexListener.php (hunk 3 above)
Comment (Toflar) [reply core-bundle/src/EventListener/SearchIndexListener.php]: I was thinking about this too but I'm not sure. Right now we don't have any message either so I kept it like that for now.

File: core-bundle/tests/DependencyInjection/ContaoCoreExtensionTest.php (hunk 4 above)
Comment (Toflar) [reply core-bundle/tests/DependencyInjection/ContaoCoreExtensionTest.php:1979]: Not really no. This is not testing the real container but only a mock. So we wouldn't test anything at all.

**Comment by Toflar** [comment]
All comments addressed, merged master into the PR. All green 🎉

File: core-bundle/src/Search/Escargot/Subscriber/EscargotSubscriber.php [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Search\Escargot\Subscriber;
+
+use Terminal42\Escargot\Subscriber\SubscriberInterface;
+
+interface EscargotSubscriber extends SubscriberInterface
```
Comment (leofeyer) [inline core-bundle/src/Search/Escargot/Subscriber/EscargotSubscriber.php]: I just noticed that this does not follow our naming conventions. It should be `EscargotSubscriberInterface`.

**Review by leofeyer — CHANGES_REQUESTED** [review]
Besides the wrong class name, everything looks RTM to me.

File: core-bundle/src/Search/Escargot/Subscriber/EscargotSubscriber.php (hunk 5 above)
Comment (Toflar) [reply core-bundle/src/Search/Escargot/Subscriber/EscargotSubscriber.php]: Oh, indeed! Good catch! Fixed in e591905a87c3c75dbfc378a980d4920bf595895a.

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @Toflar. This is huge! 💪

**Pull request merged**